### PR TITLE
Refact/slots

### PR DESCRIPTION
**Pull request by antmat**
This PR introduces separated askPlans(slots) for hub with eventual announce/deannounce to market.
Now only slots which has enough resources on miners make announces on market. When resources are consumed and exhausted deannounce is made.
TODO: Proper tests
@3Hren @sshaman1101 @screwyprof PTAL

### CODE REVIEW

File: insonmnia/hub/slot.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	"github.com/sonm-io/core/util"
+	"go.uber.org/zap"
+)
+
+type AskPlan struct {
+	Id    string
+	Order *structs.Order
+}
+
+type AskPlansData map[string]*AskPlan
```
Comment (screwyprof) [inline insonmnia/hub/slot.go]: Even though Go looks a bit like C, its compiler works differently. And C analogy does not always work with Go. Passing by value in Go may be significantly cheaper than passing by pointer. This happens because Go uses escape analysis to determine if variable can be safely allocated on function's stack frame, which could be much cheaper then allocating variable on the heap. Passing by value simplifies escape analysis in Go and gives variable a better chance to be allocated on the stack.

In other words, it's not really necessary to use a pointer to AskPlan in this case.
I've made an example (even though it's quite primitive, it shows the case) 
https://paste.ofcode.org/tD3nMDxewkYk2ZrVc9eipY

```
❯ go test -bench=.
goos: darwin
goarch: amd64
pkg: github.com/screwyprof/refvalue
BenchmarkByRef-8   	300000000	         5.47 ns/op
BenchmarkByVal-8   	300000000	         5.37 ns/op
PASS
ok  	github.com/screwyprof/refvalue	4.355s
```

File: insonmnia/hub/server.go [hunk 2]
```diff
@@ -1081,75 +1091,49 @@ func (h *Hub) SetDeviceProperties(ctx context.Context, request *pb.SetDeviceProp
 	return &pb.Empty{}, nil
 }
 
+//TODO: It actually should be called getAskPlans
```
[the remark this answers is not in the document]
Comment (antmat) [reply insonmnia/hub/server.go]: main goal of this todo is to remove ambiguity between ask plans, slots and orders.

File: insonmnia/hub/config.go [hunk 3]
```diff
@@ -21,7 +21,8 @@ type LocatorConfig struct {
 }
 
 type MarketConfig struct {
-	Endpoint string `required:"true" yaml:"endpoint"`
+	Endpoint        string `required:"true" yaml:"endpoint"`
+	UpdatePeriodSec uint64 `default:"60" yaml:"update_period_sec"`
```
[the remark this answers is not in the document]
Comment (antmat) [reply insonmnia/hub/config.go]: How do you suggest to manage duration inside of serializable/deserializable config? I think we need ability to add some initialization code to configs, but it's definetly out of the scope.

File: insonmnia/hub/slot.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	result := make(map[string]*pb.Slot)
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	for id, plan := range a.Data {
+		result[id] = plan.Order.Slot
+	}
+	return result
+}
+
+func (a *AskPlans) Dump() AskPlansData {
```
[the remark this answers is not in the document]
Comment (antmat) [reply insonmnia/hub/slot.go]: it is consistent with https://github.com/sonm-io/core/blob/master/insonmnia/hub/orders.go#L126

File: insonmnia/hub/slot.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+			log.G(a.ctx).Warn("failed to touch orders on market", zap.Error(err))
+		}
+	}
+	if changed {
+		a.sync()
+	}
+	return nil
+}
+
+//TODO: do we need to signal about error?
```
[the remark this answers is not in the document]
Comment (antmat) [reply insonmnia/hub/slot.go]: it is private method and all it's failures are temporary in general. We do not need to make any specific action in this case

[the remark this answers is not in the document]
Comment (screwyprof) [reply insonmnia/hub/slot.go]: What about all the non-private methods? ) We have a bunch of them logging errors directly, instead of making them bubble-up for further propagation :)

[the remark this answers is not in the document]
Comment (screwyprof) [reply insonmnia/hub/slot.go]: I'm not being picky ) I'm just making notes for the latter discussion which may end up in stating all the mentioned nitty-gritties in a Code Style Guide or smth )

File: insonmnia/hub/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (screwyprof) [reply insonmnia/hub/config.go]: Got you :) Let's leave it as it is for now.

File: insonmnia/hub/slot.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (screwyprof) [reply insonmnia/hub/slot.go]: I see. It would also be great to make context in Run consistent as well :)

File: insonmnia/hub/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (oopcode) [reply insonmnia/hub/config.go]: You actually *can* use `time.Duration` and specify values as `"14h"`, `"89s"`, etc.

[the remark this answers is not in the document]
Comment (3Hren) [reply insonmnia/hub/config.go]: Vote for duration, gives more mercy for humans.

File: insonmnia/hub/slot.go (hunk 1 above)
Comment (3Hren) [reply insonmnia/hub/slot.go]: I feel it was here to allow nullable state, it's go, remember? No `Option<AskPlan>` , no `AskPlan?`... and this is sad.
However exactly in this case it doesn't matter, except implicit sharing (which doesn't seems).

Vote for removing pointer.

File: insonmnia/hub/slot.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+func (a *AskPlans) announcePlan(plan *AskPlan) {
+	createdOrder, err := a.market.CreateOrder(a.ctx, plan.Order.Unwrap())
+	if err != nil {
+		log.S(a.ctx).Warnf("failed to announce ask plan with id{} on market - {}", plan.Id, zap.Error(err))
```
[the remark this answers is not in the document]
Comment (3Hren) [reply insonmnia/hub/slot.go]: S or G, whatever IMO. Personally I think we abuse logging with attributes in many cases, because attributes are planned to be used in some further log-collecting system in the pipeline. Thus we'd better discuss about which things are fit in attributes and which are not.
